docs/GDPs/gdp-0003-duck-storage.rst

```diff
@@ -0,0 +1,750 @@
+    triggered before the next access to the host buffer.
+
+:code:`set_host_modified(self: Storage) -> None`
+    Mark the host buffer as modified, so that a copy from host to devcie is automatically triggered
+    before the next access to the device buffer.
+
+:code:`set_synchronized(self: Storage) -> None`
+    Mark host and device buffers as synchronized, meaning they are equal. (In case the user has done
+    this synchronization manually).
+
+:code:`synchronize(self: Storage) -> None`,
+    Triggers a copy between host and device buffers if the host or device, respectively are
+    marked as modified. After the call the buffers are flagged as
+    synchronized.
+
+Choosing the Device
+===================
+
+For the synchronized memory classes (be it by CUDA or by GT4Py), the the device where data is
+written in :code:`__setitem__` is chosen depending on
+
+:code:`CudaManagedGPUStorage`
+    The device is chosen to be GPU if and only if the value is GPU-enabled. A value is considered
+    GPU enabled, if:
+
+    1. It implements the :code:`__gt_data_interface__` and information for a :code:`"gpu"` buffer is
+       provided. In this case, the :code:`"acquire"` method is called before reading.
+    2. It does not implement the data interface but it is compatible with :code:`cp.asarray`, which
+       includes values implementing the :code:`__cuda_array_interface__`.
+
+:code:`SoftwareManagedGPUStorage`
+    The device is chosen to be GPU if and only if the value is considered as on GPU. If the value
+    is itself a :code:`SoftwareManagedGPUStorage`, it is considered as on GPU, if the buffers are
+    either in sync or the GPU buffer is modified. If however, the value is not a
+    :code:`SoftwareManagedGPUStorage`, the same logic applies as for the
+    :code:`CudaManagedGPUStorage`. Note that :code:`xarray` :code:`DataArray` will be treated based
+    on the data interface and not the underlying storage.
+
+.. _storage_types:
+
+Storage Types
+^^^^^^^^^^^^^
+
+GT4Py Storages objects type should be subclasses of the main :code:`Storage` class. Depending on
+the choice of the :code:`device` and :code:`managed` values (see Section :ref:`constructors`), the
+type is one of :code:`CPUStorage`, :code:`GT4PySyncedGPUStorage`, :code:`CUDASyncedGPUStorage`
+or :code:`GPUStorage`.
+
+Their purpose is as follows:
+
+:code:`CUDAManagedGPUStorage`
+    Internally holds a reference to a `NumPy <https://numpy.org/>`_ `ndarray`. The memory is however
+    allocated as CUDA unified memory, meaning that the same memory can be accessed from GPU, and
+    synchronization is taken care of by the CUDA runtime.
+
+:code:`CPUStorage`
+    It holds a reference to a `NumPy <https://numpy.org/>`_ :code:`ndarray`.
+
+:code:`GPUStorage`
+    Internally holds a reference to a `CuPy <https://cupy.chainer.org/>`_ `ndarray`. This storage
+    does not have a CPU buffer.
+
+:code:`SoftwareManagedGPUStorage`
```

Comment:
nit: The previous name of `ExplicitlyManaged` was good for this object, I think. "Software" reads like an external binary running on the host system outside of Python. It contrasts well with CUDAManagedGPUStorage.